rclcpp/test/rclcpp/test_intra_process_manager_with_allocators.cpp

```diff
@@ -0,0 +1,221 @@
+  }
+
+  void deallocate(T * ptr, size_t size)
+  {
+    (void)size;
+    if (!ptr) {
+      return;
+    }
+    num_deallocs++;
+    std::free(ptr);
+  }
+
+  template<typename U>
+  struct rebind
+  {
+    typedef MyAllocator<U> other;
+  };
+};
+
+template<typename T, typename U>
+constexpr bool operator==(
+  const MyAllocator<T> &,
+  const MyAllocator<U> &) noexcept
+{
+  return true;
+}
+
+template<typename T, typename U>
+constexpr bool operator!=(
+  const MyAllocator<T> &,
+  const MyAllocator<U> &) noexcept
+{
+  return false;
+}
+
+/*
+   This tests the case where a custom allocator is used correctly, i.e. the same
+   custom allocator on both sides.
+ */
+TEST(TestIntraProcessManagerWithAllocators, custom_allocator) {
+  auto context = std::make_shared<rclcpp::Context>();
+  context->init(0, nullptr);
+  auto node = std::make_shared<rclcpp::Node>(
+    "custom_allocator_test",
+    rclcpp::NodeOptions().context(context).use_intra_process_comms(true));
+
+  uint32_t counter = 0;
+  auto callback =
+    [&counter](std::shared_ptr<const test_msgs::msg::Empty>) {
+      ++counter;
+    };
+
+  using Alloc = MyAllocator<void>;
+  auto alloc = std::make_shared<Alloc>();
+  rclcpp::PublisherOptionsWithAllocator<Alloc> publisher_options;
+  publisher_options.allocator = alloc;
+  auto publisher =
+    node->create_publisher<test_msgs::msg::Empty>("custom_allocator_test", 10, publisher_options);
+
+  rclcpp::SubscriptionOptionsWithAllocator<Alloc> subscription_options;
+  subscription_options.allocator = alloc;
+  auto msg_mem_strat =
+    std::make_shared<
+    rclcpp::message_memory_strategy::MessageMemoryStrategy<test_msgs::msg::Empty, Alloc>
+    >(alloc);
+  auto subscriber = node->create_subscription<test_msgs::msg::Empty>(
+    "custom_allocator_test", 10, callback, subscription_options, msg_mem_strat);
+
+  using rclcpp::memory_strategies::allocator_memory_strategy::AllocatorMemoryStrategy;
+  std::shared_ptr<rclcpp::memory_strategy::MemoryStrategy> memory_strategy =
+    std::make_shared<AllocatorMemoryStrategy<Alloc>>(alloc);
+
+  rclcpp::ExecutorOptions options;
+  options.memory_strategy = memory_strategy;
+  options.context = context;
+  rclcpp::executors::SingleThreadedExecutor executor(options);
+
+  executor.add_node(node);
+
+  using MessageAllocTraits =
+    rclcpp::allocator::AllocRebind<test_msgs::msg::Empty, Alloc>;
+  using MessageAlloc = MessageAllocTraits::allocator_type;
+  using MessageDeleter = rclcpp::allocator::Deleter<MessageAlloc, test_msgs::msg::Empty>;
+  MessageDeleter message_deleter;
+  MessageAlloc message_alloc = *alloc;
+  rclcpp::allocator::set_allocator_for_deleter(&message_deleter, &message_alloc);
+
+  auto ptr = MessageAllocTraits::allocate(message_alloc, 1);
+  MessageAllocTraits::construct(message_alloc, ptr);
+  std::unique_ptr<test_msgs::msg::Empty, MessageDeleter> msg(ptr, message_deleter);
+  EXPECT_NO_THROW(
+  {
+    publisher->publish(std::move(msg));
+    executor.spin_some();
```

Comment:
@wjwwood shall we put an expectation on `counter`? Like:

```
EXPECT_EQ(1u, counter);
```

Reply:
Also, an expectation on the message address would be nice too (to make sure the custom allocator didn't induce a copy).

Reply:
Well, that was more complicated than I expected, but I did it 060526d :)